### DISCUSSION TITLE
last.fm "similar artist" script

### DISCUSSION THREAD

**Discussion by td598** [jcorporation/mympd-scripts#31] [Q&A]
I try to setup a script to get similar artists from last.fm. I get the following answer when I enter the appropriate URL from last-fm API in my explorer:
"<lfm status="ok">
<similarartists artist="Electric Light Orchestra">
<artist>
<name>Jeff Lynne</name>
<mbid>372304d0-5e6a-4030-bc2b-09bdb0ee8abb</mbid>
<match>1</match>
<url>https://www.last.fm/music/Jeff+Lynne</url>
<image size="small">
https://lastfm.freetls.fastly.net/i/u/34s/2a96cbd8b46e442fc41c2b86b821562f.png
</image>
<image size="medium">
https://lastfm.freetls.fastly.net/i/u/64s/2a96cbd8b46e442fc41c2b86b821562f.png
</image>
<image size="large">
https://lastfm.freetls.fastly.net/i/u/174s/2a96cbd8b46e442fc41c2b86b821562f.png
</image>
<image size="extralarge">
https://lastfm.freetls.fastly.net/i/u/300x300/2a96cbd8b46e442fc41c2b86b821562f.png
</image>
<image size="mega">
https://lastfm.freetls.fastly.net/i/u/300x300/2a96cbd8b46e442fc41c2b86b821562f.png
</image>
<image size="">
https://lastfm.freetls.fastly.net/i/u/300x300/2a96cbd8b46e442fc41c2b86b821562f.png
</image>
<streamable>0</streamable>
</artist>
<artist>
<name>Supertramp</name>
<mbid>d2ff6b6b-fc30-48dc-8952-06f9d8fc64f8</mbid>
<match>0.648610</match>
<url>https://www.last.fm/music/Supertramp</url>
<image size="small">
https://lastfm.freetls.fastly.net/i/u/34s/2a96cbd8b46e442fc41c2b86b821562f.png
</image>
<image size="medium">
https://lastfm.freetls.fastly.net/i/u/64s/2a96cbd8b46e442fc41c2b86b821562f.png
</image>
<image size="large">
https://lastfm.freetls.fastly.net/i/u/174s/2a96cbd8b46e442fc41c2b86b821562f.png
</image>
<image size="extralarge">
https://lastfm.freetls.fastly.net/i/u/300x300/2a96cbd8b46e442fc41c2b86b821562f.png
</image>
<image size="mega">
https://lastfm.freetls.fastly.net/i/u/300x300/2a96cbd8b46e442fc41c2b86b821562f.png
</image>
<image size="">
https://lastfm.freetls.fastly.net/i/u/300x300/2a96cbd8b46e442fc41c2b86b821562f.png
</image>
<streamable>0</streamable>
</artist>"

I think I must use the following function to read this information in myMPD:

"rc, code, headers, body = mympd.http_client("GET", url, "", "")"

But I can not get it to work properly. I get "400" back. I assume it has someting to do with the "extra_headers" that need to be provided?

**Comment by jcorporation** [comment]
You should ask questions in Discussions. Issues are only for bug reports.

**Comment by jcorporation** [comment]
Please post the exact URL you are calling.

**Comment by td598** [comment]
I used the one from the last.fm web site:
https://ws.audioscrobbler.com/2.0/?method=artist.getsimilar&artist=cher&api_key=YOUR_API_KEY&format=json
I used the above one and the same but without the "format=json" parameter. I registered in last.fm and used my own "MY_API_KEY".

**Comment by jcorporation** [comment]
This works for me:

```lua
local uri = "https://ws.audioscrobbler.com/2.0/?method=artist.getsimilar&artist=cher&api_key=xxxx&format=json"
local rc, code, headers, body = mympd.http_client("GET", uri, "", "")
```

**Comment by td598** [comment]
I used "url" as local parameter. That seems not to work. When I chaneg to uri or something else it works. I am new to this (obviously) and try to learn. I really like the way myMPD is setup up. It allows a lot of flexibility. Thank you very much!

**Reply by jcorporation** [reply]
You can name the variable whatever you want, as long you reference it correctly, it is only a name.

**Comment by td598** [comment]
I got it to work but now it would be great to get the list of similar artists on the "actual playing" page as it is possible for the song texts. Is there a way?

**Reply by jcorporation** [reply]
Currently there is no way to display it on the "playback" page but it can be displayed on the home screen. You must create a home widget.

https://jcorporation.github.io/myMPD/050-scripting/widgets/#common-javascript-functions

Would you want to share your code?

**Reply by td598** [reply]
Sure I will share my code. Is there a way to call a web-site outside of myMPD. I would like to link to the last.fm page of the similar artist similar as in the WidgetAlbums script when an Album is called.

**Reply by jcorporation** [reply]
You can return a html link.

**Comment by td598** [comment]
This is how I have set it up:
   \" local url_ = "<url class=\"list-group-item list-group-item-action clickable\"><a href=" .. url_str .. ">" .. name_str .. "</a></url>"\"
The idea is to show the name of the artist and when clicking go to the respective web page. It shows the the right URL but it does not open the page.

**Reply by jcorporation** [reply]
This is invalid html. `url` is no valid html tag.

You can find the documentaion for Bootstrap List groups here: https://getbootstrap.com/docs/5.3/components/list-group/#for-links-and-buttons

```
<a href="LINK" target="_blank" class="list-group-item list-group-item-action">TEXT</a>
```

Replace `LINK` with the link target and TEXT with the text that show up.

**Comment by td598** [comment]
This works only if I click on the pop-up menue that opens when I click on an artist and select "open in a separate tab". Nevertheless thank you. I like this gimmik.

Still I think I do not understand how to declare variable right. I can not use "num_artists"  as parameter for "limit" in the url for the get call but have to state a fix number? 

This is my code:

local headers_ = "Content-type: text/html\r\n"
local line = {}
local num_artists = 10
local method, limit, api_key, artist, song, x
local j, k, l, m, n, rc, code
local rows_ = {}
local name_str
local url_str

mympd.init()

if mympd_arguments.count ~= nil then
    num_artists = tonumber(mympd_arguments.song_count)
end

if mympd_state.current_song == nil then
  local body = "<div class=\"list-group\">".."__no song__" .."</div>"
  return mympd.http_reply("200", headers_, body)
else
  local uri_t = mympd_state.current_song.uri
  rc, song = mympd.api("MYMPD_API_SONG_DETAILS", {uri = uri_t})
end

local artist_ = song.Artist[1]
local artist_str = string.gsub(artist_, " ", "%%20")

local url_ = string.format("https://ws.audioscrobbler.com/2.0/?method=artist.getsimilar&limit=10&artist=%s&api_key=YOUR LAST.FM KEY",artist_str)
local headers, body_
rc, code, headers, body_ = mympd.http_client("GET", url_, "", "")

j = 1
if rc == 0 and code == 200 then
  for x in body_:gmatch("[^\r\n]+") do
    line[j] = x
    j = j + 1
  end
else
   return "error"
end

j = 1
while line[j] do
  k, l = string.find(line[j], "<name>")
  if k ~= nil then
     m, n = string.find(line[j], "</name>")
     name_str = string.sub(line[j], l+1, m-1)
  else
  if string.find(line[j], "<url>") then
     local url_str_h = string.gsub(line[j], "<url>", "")
     url_str = string.gsub(url_str_h, "</url>", "")
  else
  if string.find(line[j], "</artist>") then
    local url_ = "<a href="..url_str.." target=\"_blank\" class=\"list-group-item list-group-item-action\">" .. name_str .. "</a>"
    table.insert(rows_, url_)
  end end end
  j = j + 1
end

local url_ = "<a class=\"list-group-item\">" .. "for " .. song.Artist[1] .. " ->" .. "</a>"
table.insert(rows_, 1, url_)

local body = "<div class=\"list-group list-group-flush\">" .. table.concat(rows_) .. "</div>"

return mympd.http_reply("200", headers_, body)
